site/content/posts/2021-11-09-E2E-Skeleton-Plugin.md

```diff
@@ -0,0 +1,135 @@
+date: 2021-11-09
+slug: plugin-starter
+---
+
+## Introduction
+
+When you decide to test your Kubernetes native code it can be a big chore. When I've been in that position, I do one of the following:
+ - start a new project
+   - Simple in concept but you're left with a ton of boilerplate code so that you can do simple things like
+ check if your app is running. You've got to deal with setting up client-go, you've got to figure out all the wiring
+ to set up builds, images, API clients, avoid test collision in the cluster, etc.
+ - vendor upstream k8s testing code
+   - This has the benefit of giving you lots of helper functions, but it is a huge dependency and bloats your tests. In addition,
+ even with all the new helpers, there is SO much code that it can be confusing about what to use and how to use it.
+
+Instead, we want to suggest you try our new [e2e-skeleton plugin][skeleton]. It is obviously just a starting point (not a plugin you would
+run without modification), but it handles much of the basic boilerplate code for you, including:
+ - Configures an API client for k8s CRUD operations
+ - Reports results in a format that Sonobuoy understands (e.g. `sonobuoy results` would be able to tell you pass/fail/etc)
+ - Reports results to the Sonobuoy aggregator once done
+ - Creates/destroys namespaces for each test to avoid collisions
+ - Sends progress updates to the Sonobuoy aggregator so that `sonobuoy status` can report on test progress while running
+
+In this blog post I'll show off the plugin, its features, and show you how to skip the boilerplate so you can get to testing your business logic.
+
+## How to use the e2e-skeleton plugin
+
+This is a "plugin" which is _meant_ to be modified prior to use. Currently, it is
+just a skeleton for you to put your tests into. So in order to get set up and write your own tests,
+you have to:
+
+ - Clone the repo (or copy the example code) `git clone https://github.com/vmware-tanzu/sonobuoy-plugins`
+ - Modify `build.sh` to contain your own registry/image name
+ - Modify `plugin.yaml` to reference the registry/image for your tests
+ - Add more tests!
+
+## Integrations with Sonobuoy
+
+#### Preconfigured Project Files
+ 
+Since this code is meant to be your starting point for your own plugin, we
+include the basic files you'll need during your plugin's lifecycle:
+
+- Dockerfile
+  - All Sonobuoy plugins end up launching containers in your Kubernetes cluster. So to make a plugin you'll need to containerize your code. We've done that for you already and used a base image that was easy to work with and debug. Feel free to modify this as desired.
+- Instructions/script to build your code
+  - You've got the code and Dockerfile, but you still have to actually build/push the container. We provide a simple `build.sh` script that will build/push the image to your specified registry. (This is currently a very minimal starting point and would love help making it more robust.)
```

Comment:
`(This is currently a very minimal starting point and would...` > `(This is currently a very minimal starting point and we would...`